### PR TITLE
Core addons tests for multi node cluster

### PR DESCRIPTION
**Pull request by sachinkumarsingh092**
`microk8s.reset` freezes a multinode cluster setup when testing on a multinode setup. This PR introduces the changes to allow these tests to run on a multi-node cluster. 

Changes:
- Introduce nodeAffinity for hostpath storage
- Fix certificate tests
- Fix DNS addon enablement for version above 1.27

### CODE REVIEW

File: .github/workflows/run-tests.yml [hunk 1]
```diff
@@ -24,6 +24,7 @@ jobs:
           sudo apt-get install python3-setuptools
           sudo pip3 install --upgrade pip
           sudo pip3 install -U pytest
+
```
Comment (ktsakalozos) [inline .github/workflows/run-tests.yml:27]: You do not need this empty line

File: addons/hostpath-storage/storage.yaml [hunk 2]
```diff
@@ -17,6 +17,15 @@ spec:
         k8s-app: hostpath-provisioner
     spec:
       serviceAccountName: microk8s-hostpath
+      affinity:
```
Comment (ktsakalozos) [inline addons/hostpath-storage/storage.yaml:20]: What is the purpose of this affinity rule? This "hostpath-test-node" does not look right.

File: tests/templates/ingress.yaml [hunk 3]
```diff
@@ -18,7 +18,7 @@ spec:
         app: microbot
     spec:
       containers:
-        - image: cdkbot/microbot-$ARCH
+        - image: cdkbot/microbot-amd64
```
Comment (ktsakalozos) [inline tests/templates/ingress.yaml:21]: You do not want to commit this.

File: tests/test-addons.py [hunk 4]
```diff
@@ -57,7 +63,22 @@ def test_help_text(self):
...
+        cmd = "/snap/bin/microk8s.version"
+        version_str = run_until_success(cmd, timeout_insec=300)
+        version = check_output(
+            f"echo {version_str} | sudo grep -oP '\\d+\\.\\d+' ",
+            shell=True,
+            encoding="utf-8",
+        ).strip()
+        minor_version = int(version[2:])
+
+        # dns is enabled by default in versions 1.27+
```
Comment (ktsakalozos) [inline tests/test-addons.py:77]: I am not sure what is the point of this given the master branch of the addons repository is used on the 1.27 (for now) MicroK8s only. If we want to handle the test behavior on pre-1.27 releases we should be patching the respective branches (eg 1.26).

File: tests/test-addons.py [hunk 5]
```diff
@@ -231,7 +268,6 @@ def test_observability(self):
         validate_observability()
         print("Disabling observability")
         microk8s_disable("observability")
-        microk8s_reset()
```
Comment (ktsakalozos) [inline tests/test-addons.py:234]: Why was this removed? I thought the reset is made a no-op in multinode tests.

File: tests/templates/pvc-pvdir.yaml [hunk 6]
```diff
@@ -27,6 +27,15 @@ apiVersion: v1
 metadata:
   name: hostpath-test-pod-pvdir
 spec:
+  affinity:
```
Comment (ktsakalozos) [inline tests/templates/pvc-pvdir.yaml:30]: Why is this affinity rule needed? Are we saying that in a multinode cluster hostpath-provider will not work?

File: tests/templates/cert-manager-aio-test.yaml [hunk 7]
```diff
@@ -42,6 +42,8 @@ metadata:
     app: microbot
   name: microbot
 spec:
+  externalTrafficPolicy: Cluster
+  type: LoadBalancer
```
Comment (ktsakalozos) [inline tests/templates/cert-manager-aio-test.yaml:46]: This says we will need a loadbalancer to get this service up. What is the error we get if we do not do this? Does the test that check the cert-manager install metallb?

File: .github/workflows/run-tests.yml (hunk 1 above)
Comment (sachinkumarsingh092) [reply .github/workflows/run-tests.yml:27]: Removed.

File: addons/hostpath-storage/storage.yaml (hunk 2 above)
Comment (sachinkumarsingh092) [reply addons/hostpath-storage/storage.yaml:20]: In multinode setup, we want the hostpath pod to be on the same node as the node from which the tests are executed, other wise `validate_storage()` and `validate_storage_custom_pvdir()` fail if the hostpath pod is created to some other node.

File: tests/test-addons.py (hunk 4 above)
Comment (sachinkumarsingh092) [reply tests/test-addons.py:77]: Okay removing from this PR.

File: tests/test-addons.py (hunk 5 above)
Comment (sachinkumarsingh092) [reply tests/test-addons.py:234]: In the `clean_up()` fixture we check the setup. If it's a multinode, we skip `microk8s_reset` other wise we execute it at the end of every test.

File: tests/templates/pvc-pvdir.yaml (hunk 6 above)
Comment (sachinkumarsingh092) [reply tests/templates/pvc-pvdir.yaml:30]: This is for the same reason as above, to ensure that that this pod is on the same node as that of hostpath pod.

File: addons/hostpath-storage/storage.yaml (hunk 2 above)
Comment (ktsakalozos) [reply addons/hostpath-storage/storage.yaml:20]: The `storage.yaml` is what all users deploy. This affinity rule is only for our own tests. We should not be applying it to all deployments.

File: tests/templates/cert-manager-aio-test.yaml (hunk 7 above)
Comment (sachinkumarsingh092) [reply tests/templates/cert-manager-aio-test.yaml:46]: We were not able to finish the challenge (`Waiting for http-01 challenge propagation`) and it was a continuous error. This was probably caused due to non-resolution of the host `microbot.100.100.100.100.nip.io`.

File: addons/hostpath-storage/storage.yaml (hunk 2 above)
Comment (ktsakalozos) [reply addons/hostpath-storage/storage.yaml:20]: Could we inject any affinity rules at the time of the test?

Comment (ktsakalozos) [reply addons/hostpath-storage/storage.yaml:20]: Ok, lets do that.

Comment (sachinkumarsingh092) [reply addons/hostpath-storage/storage.yaml]: Done.

File: tests/test-addons.py [hunk 8]
```diff
@@ -110,6 +116,12 @@ def test_basic(self):
         Sets up and tests dashboard, dns, storage, registry, ingress, metrics server.
 
         """
+        # Set labels
+        node_name = kubectl(f"get node -o jsonpath='{{.items[0].metadata.name}}'")
```
[the remark this answers is not in the document]
Comment (sachinkumarsingh092) [reply tests/test-addons.py:120]: Oh yes. It can easily be a false positive. I was just getting the right name so didn't bother.
A cleaner approach can be reading the name from `/etc/hostname`. I will do that.
